src/include/spinlock.h

```diff
@@ -0,0 +1,103 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2017 SUSE LINUX GmbH
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+*/
+
+#ifndef CEPH_SPINLOCK_HPP
+#define CEPH_SPINLOCK_HPP 1
+
+#include <atomic>
+
+namespace ceph {
+
+class Spinlock;
+
+inline void spin_lock(std::atomic_flag& lock);
+inline void spin_unlock(std::atomic_flag& lock);
+inline void spin_lock(ceph::Spinlock& lock);
+inline void spin_unlock(ceph::Spinlock& lock);
+
+class Spinlock {
+  mutable std::atomic_flag spinlock { false };
+
+public:
+  /// acquire spinlock
+  void lock() const {
+    ceph::spin_lock(spinlock);
+  }
+  /// release spinlock
```

Comment:
It just tells the compiler and reader that this member function causes no side-effects on the object itself. In addition to expressing intent, it can enable some extra optimizations.